README.txt
*******************************
* LibRepository 0.2.0         *
*******************************

30 May 2008

1. INTRODUCTION
---------------
LibRepository provides a simple abstraction layer to access bulk content that
is organized in a hierarchical layer.

Unlike the JSR-000170, this library does not aim to solve all problems
associated with content storages. The main purpose of LibRepository is to
give users an abstract view over an filesystem like structure so that content
generator and content consumer do no longer have to make assumptions about
where to store the generated content.

The repositories described here should not be used to store other things than
BLOBs.

For the latest news and information about LibRepository, please refer to:

    http://reporting.pentaho.org/librepository/


2. SUPPORT
----------
Free support is available via the Pentaho Reporting forum, follow the link
from the LibRepository home page.  Please note that questions are
answered by volunteers, so there is no guaranteed response time or
level of service.

Please avoid e-mailing the developers directly for support questions.
If you post a message in the forum, then everyone can see the
question, and everyone can see the answer.


3. REPORTING BUGS
-----------------
If you find bugs in LibRepository, we'd like to hear about it so that we
can improve future releases of LibRepository.  Please post a bug report
in our JIRA bug-tracking system at:

    http://jira.pentaho.org/browse/PRE

Please be sure to provide as much information as you can.  We need to
know the version of LibRepository that you are using, the JDK version,
and the steps required to replicate the bug.  Include any other
information that you think is relevant.


4. ANT
------
We use an open source build tool called Ant to build LibRepository.  An
Ant script (tested using Ant 1.6.2) is included in the distribution:

    <librepository-directory>/build.xml

You can find out more about Ant at:

    http://ant.apache.org/

Ant is licensed under the terms of the Apache Software License (a
popular open source software license).


5. OTHER FEEDBACK
-----------------
For other feedback and comments, please post a message on the
Pentaho forums. The Forum is available at

  http://forums.pentaho.org/



ChangeLog.txt
---------------
1.  WHAT's NEW
---------------
A list of changes in recent versions:

0.2.0: (30-May-2008)
       * Code-hardening: Explict checks for Null-Values added.
  
       * Switched from JCommon to LibBase. All version information is now contained in
         the manifest. The Project-Info implementation reads the version numbers from
         the Jar's Manifest or the compiler-output-directory.

       * Completed the ZIP repositories. The library now provides three flavors:
         - read-only streaming repository
         - write-only streaming repository
         - read-write memory-buffered repository

       * [BUG] Generating ZIP repositories was generally unstable. The repository did
         not write un-compressed entries as it required a precomputed CRC32 checksum.
         
0.1.6: (30-Nov-2007)
       * Updated the error messages for the file-repository.
         We now explicitly check whether the directories for
         the content-locations exist. If they do not exist,
         the implementation throws exceptions ..

0.1.5: (02-Nov-2007)
        * Upgraded to jcommon-1.0.12

0.1.4: (16-Oct-2007)
        * Automatic fixes as reported by IntelliJ-IDEA's inspections
    
        * Updated the copyright header on all java-files

0.1.3: (24-Sep-2007)
        * All releases are now built with JDK 1.4. However, we are still compatible
          with JDK 1.2.2 and all releases run and compile under that release.

0.1.2: (30-Jun-2007)
        * [BUG] The ZIP-Repository was not closed correctly, causing unreadable files.
    
        * The default-name-generator tries to extract a suitable default-suffix from
          the given default-filename.

0.1.1: (27-May-2007)
        * Added a mime-type mapping for CSS-files. Mime-Mapping should be configurable.
    
0.1.0: (03-Dec-2006)
        * Initial version.


source/org/pentaho/reporting/libraries/repository/zip/ZipRepository.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.repository.zip;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.Deflater;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

import org.pentaho.reporting.libraries.repository.ContentIOException;
import org.pentaho.reporting.libraries.repository.ContentLocation;
import org.pentaho.reporting.libraries.repository.DefaultMimeRegistry;
import org.pentaho.reporting.libraries.repository.MimeRegistry;
import org.pentaho.reporting.libraries.repository.Repository;
import org.pentaho.reporting.libraries.repository.RepositoryUtilities;
import org.pentaho.reporting.libraries.base.util.IOUtils;

/**
 * A read-write repository based on ZIP streams. The repository can be created using a existing
 * zip file as initial content. The repository will be fully buffered, so nothing is written until
 * the whole repository is closed. For a streaming solution use the zipwriter-repository instead.
 *
 * @author Thomas Morgner
 */
public class ZipRepository implements Repository
{
  private ZipContentLocation root;
  private MimeRegistry mimeRegistry;

  public ZipRepository()
  {
    this(new DefaultMimeRegistry());
  }

  public ZipRepository(final MimeRegistry mimeRegistry)
  {
    if (mimeRegistry == null)
    {
      throw new NullPointerException();
    }

    this.mimeRegistry = mimeRegistry;
    this.root = new ZipContentLocation(this, null, "");
  }

  public ZipRepository(final InputStream in) throws IOException
  {
    this(in, new DefaultMimeRegistry());
  }

  public ZipRepository(final InputStream in, final MimeRegistry mimeRegistry) throws IOException
  {
    this(mimeRegistry);

    final ZipInputStream zipIn = new ZipInputStream(in);
    ZipEntry nextEntry = zipIn.getNextEntry();
    while (nextEntry != null)
    {
      final String[] buildName = RepositoryUtilities.splitPath(nextEntry.getName(), "/");
      if (nextEntry.isDirectory())
      {
        root.updateDirectoryEntry(buildName, 0, nextEntry);
      }
      else
      {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        DeflaterOutputStream dos = new DeflaterOutputStream(bos, new Deflater(nextEntry.getMethod()));
        IOUtils.getInstance().copyStreams(zipIn, dos);
        dos.flush();
        dos.close();

        root.updateEntry(buildName, 0, nextEntry, bos.toByteArray());
      }

      zipIn.closeEntry();
      nextEntry = zipIn.getNextEntry();
    }
  }

  public ContentLocation getRoot() throws ContentIOException
  {
    return root;
  }

  public MimeRegistry getMimeRegistry()
  {
    return mimeRegistry;
  }

  public void write(final OutputStream outputStream) throws IOException, ContentIOException
  {
    RepositoryUtilities.writeAsZip(outputStream, this);
  }

  public void writeToZipStream(final ZipOutputStream zipOutputStream,
                               final Repository repository) throws IOException, ContentIOException
  {
    RepositoryUtilities.writeToZipStream(zipOutputStream, repository);
  }


}


source/org/pentaho/reporting/libraries/repository/file/FileContentLocation.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.repository.file;

import java.io.File;
import java.io.IOException;

import org.pentaho.reporting.libraries.repository.ContentCreationException;
import org.pentaho.reporting.libraries.repository.ContentEntity;
import org.pentaho.reporting.libraries.repository.ContentIOException;
import org.pentaho.reporting.libraries.repository.ContentItem;
import org.pentaho.reporting.libraries.repository.ContentLocation;
import org.pentaho.reporting.libraries.repository.Repository;
import org.pentaho.reporting.libraries.repository.RepositoryUtilities;
import org.pentaho.reporting.libraries.base.util.IOUtils;

/**
 * A content-location that uses a directory as backend.
 *
 * @author Thomas Morgner
 */
public class FileContentLocation extends FileContentEntity implements ContentLocation
{
  private static final long serialVersionUID = -5452372293937107734L;

  /**
   * Creates a new location for the given parent and directory.
   *
   * @param parent the parent location.
   * @param backend the backend.
   * @throws ContentIOException if an error occured or the file did not point to a directory.
   */
  public FileContentLocation(final ContentLocation parent, final File backend) throws ContentIOException
  {
    super(parent, backend);
    if (backend.exists() == false || backend.isDirectory() == false)
    {
      throw new ContentIOException("The given backend-file is not a directory.");
    }
  }

  /**
   * Creates a new root-location for the given repository and directory.
   *
   * @param repository the repository for which a location should be created.
   * @param backend the backend.
   * @throws ContentIOException if an error occured or the file did not point to a directory.
   */
  public FileContentLocation(final Repository repository, final File backend) throws ContentIOException
  {
    super(repository, backend);
    if (backend.exists() == false || backend.isDirectory() == false)
    {
      throw new ContentIOException("The given backend-file is not a directory.");
    }
  }

  /**
   * Lists all content entities stored in this content-location. This method filters out all files that have an
   * invalid name (according to the repository rules).
   *
   * @return the content entities for this location.
   * @throws ContentIOException if an repository error occured.
   */
  public ContentEntity[] listContents() throws ContentIOException
  {
    final File file = getBackend();
    final File[] files = file.listFiles();
    final ContentEntity[] entities = new ContentEntity[files.length];
    for (int i = 0; i < files.length; i++)
    {
      final File child = files[i];
      if (RepositoryUtilities.isInvalidPathName(child.getName()))
      {
        continue;
      }

      if (child.isDirectory())
      {
        entities[i] = new FileContentLocation(this, child);
      }
      else if (child.isFile())
      {
        entities[i] = new FileContentLocation(this, child);
      }
    }
    return entities;
  }

  /**
   * Returns the content entity with the given name. If the entity does not exist, an Exception will be raised.
   *
   * @param name the name of the entity to be retrieved.
   * @return the content entity for this name, never null.
   * @throws ContentIOException if an repository error occured.
   */
  public ContentEntity getEntry(final String name) throws ContentIOException
  {
    if (RepositoryUtilities.isInvalidPathName(name))
    {
      throw new IllegalArgumentException("The name given is not valid.");
    }

    final File file = getBackend();
    final File child = new File(file, name);
    if (child.exists() == false)
    {
      throw new ContentIOException("Not found:" + child);
    }
    try
    {
      if (IOUtils.getInstance().isSubDirectory(file, child) == false)
      {
        throw new ContentIOException("The given entry does not point to a sub-directory of this content-location");
      }
    }
    catch (IOException e)
    {
      throw new ContentIOException("IO Error.", e);
    }

    if (child.isDirectory())
    {
      return new FileContentLocation(this, child);
    }
    else if (child.isFile())
    {
      return new FileContentItem(this, child);
    }
    else
    {
      throw new ContentIOException("Not File nor directory.");
    }
  }

  /**
   * Creates a new data item in the current location. This method must never
   * return null. This method will fail if an entity with the same name exists in this location.
   *
   * @param name the name of the new entity.
   * @return the newly created entity, never null.
   * @throws ContentCreationException if the item could not be created.
   */
  public ContentItem createItem(final String name) throws ContentCreationException
  {
    if (RepositoryUtilities.isInvalidPathName(name))
    {
      throw new IllegalArgumentException("The name given is not valid.");
    }

    final File file = getBackend();
    final File child = new File(file, name);
    if (child.exists())
    {
      if (child.length() == 0)
      {
        // probably one of the temp files created by the pentaho-system
        return new FileContentItem(this, child);
      }
      throw new ContentCreationException("File already exists: " + child);
    }
    try
    {
      if (child.createNewFile() == false)
      {
        throw new ContentCreationException("Unable to create the file.");
      }
      return new FileContentItem(this, child);
    }
    catch (IOException e)
    {
      throw new ContentCreationException("IOError while create", e);
    }
  }

  /**
   * Creates a new content location in the current location. This method must never
   * return null. This method will fail if an entity with the same name exists in this location.
   *
   * @param name the name of the new entity.
   * @return the newly created entity, never null.
   * @throws ContentCreationException if the item could not be created.
   */
  public ContentLocation createLocation(final String name)
      throws ContentCreationException
  {
    if (RepositoryUtilities.isInvalidPathName(name))
    {
      throw new IllegalArgumentException("The name given is not valid.");
    }

    final File file = getBackend();
    final File child = new File(file, name);
    if (child.exists())
    {
      throw new ContentCreationException("Directory already exists.");
    }
    if (child.mkdir() == false)
    {
      throw new ContentCreationException("Unable to create the directory");
    }
    try
    {
      return new FileContentLocation(this, child);
    }
    catch (ContentIOException e)
    {
      throw new ContentCreationException("Failed to create the content-location", e);
    }
  }

  /**
   * Checks, whether an content entity with the given name exists in this content location. This method
   * will report invalid filenames as non-existent.
   *
   * @param name the name of the new entity.
   * @return true, if an entity exists with this name, false otherwise.
   */
  public boolean exists(final String name)
  {
    if (RepositoryUtilities.isInvalidPathName(name))
    {
      return false;
    }

    final File file = getBackend();
    final File child = new File(file, name);
    return (child.exists());
  }
}


source/org/pentaho/reporting/libraries/repository/zipreader/ZipReadRepository.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.repository.zipreader;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.zip.Deflater;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import org.pentaho.reporting.libraries.repository.ContentIOException;
import org.pentaho.reporting.libraries.repository.ContentLocation;
import org.pentaho.reporting.libraries.repository.DefaultMimeRegistry;
import org.pentaho.reporting.libraries.repository.MimeRegistry;
import org.pentaho.reporting.libraries.repository.Repository;
import org.pentaho.reporting.libraries.repository.RepositoryUtilities;
import org.pentaho.reporting.libraries.base.util.IOUtils;

/**
 * Provides read-only access to ZIP files. The whole zip-file is cached in memory so this input
 * method will fail badly on huge zuip-files.
 *
 * @author Thomas Morgner
 */
public class ZipReadRepository implements Repository
{
  private ZipReadContentLocation root;
  private MimeRegistry mimeRegistry;

  public ZipReadRepository(final InputStream in) throws IOException
  {
    this(in, new DefaultMimeRegistry());
  }

  public ZipReadRepository(final InputStream in, final MimeRegistry mimeRegistry) throws IOException
  {
    this.mimeRegistry = mimeRegistry;
    root = new ZipReadContentLocation(this, null, "");

    final ZipInputStream zipIn = new ZipInputStream(in);
    ZipEntry nextEntry = zipIn.getNextEntry();
    if (nextEntry == null)
    {
      throw new IOException("This repository is empty or does not point to a ZIP file");
    }

    while (nextEntry != null)
    {
      final String[] buildName = RepositoryUtilities.splitPath(nextEntry.getName(), "/");
      if (nextEntry.isDirectory())
      {
        root.updateDirectoryEntry(buildName, 0, nextEntry);
      }
      else
      {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        DeflaterOutputStream dos = new DeflaterOutputStream(bos, new Deflater(nextEntry.getMethod()));
        IOUtils.getInstance().copyStreams(zipIn, dos);
        dos.flush();
        dos.close();

        root.updateEntry(buildName, 0, nextEntry, bos.toByteArray());
      }

      zipIn.closeEntry();
      nextEntry = zipIn.getNextEntry();
    }
  }


  public ContentLocation getRoot() throws ContentIOException
  {
    return root;
  }

  public MimeRegistry getMimeRegistry()
  {
    return mimeRegistry;
  }
}


source/org/pentaho/reporting/libraries/repository/zipwriter/ZipEntryOutputStream.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.repository.zipwriter;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Date;
import java.util.zip.CRC32;
import java.util.zip.Deflater;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.InflaterInputStream;
import java.util.zip.ZipEntry;

import org.pentaho.reporting.libraries.repository.LibRepositoryBoot;
import org.pentaho.reporting.libraries.repository.RepositoryUtilities;

/**
 * A fully buffered stream.
 *
 * @author Thomas Morgner
 */
public class ZipEntryOutputStream extends OutputStream
{
  private ByteArrayOutputStream outputStream;
  private DeflaterOutputStream deflaterOutputStream;
  private boolean closed;
  private ZipContentItem item;
  private CRC32 crc32;
  private long size;

  public ZipEntryOutputStream(final ZipContentItem item)
  {
    if (item == null)
    {
      throw new NullPointerException();
    }
    this.item = item;
    this.outputStream = new ByteArrayOutputStream();
    this.deflaterOutputStream = new DeflaterOutputStream
        (outputStream, new Deflater(RepositoryUtilities.getZipLevel(item)));
    this.crc32 = new CRC32();
    this.size = 0;
  }

  public void write(final int b)
      throws IOException
  {
    if (closed)
    {
      throw new IOException("Already closed");
    }
    deflaterOutputStream.write(b);
    crc32.update(b);
    size += 1;
  }

  public void write(final byte[] b, final int off, final int len)
      throws IOException
  {
    if (closed)
    {
      throw new IOException("Already closed");
    }
    deflaterOutputStream.write(b, off, len);
    crc32.update(b, off, len);
    size += len;
  }

  public void close()
      throws IOException
  {
    if (closed)
    {
      // A duplicate close is just a NO-OP as with all other output streams.
      return;
    }

    deflaterOutputStream.close();
    final byte[] data = outputStream.toByteArray();
    final ByteArrayInputStream bin = new ByteArrayInputStream(data);
    final InflaterInputStream infi = new InflaterInputStream(bin);

    final ZipRepository repository = (ZipRepository) item.getRepository();

    final String contentId = (String) item.getContentId();
    final ZipEntry zipEntry = new ZipEntry(contentId);

    final Object comment = item.getAttribute(LibRepositoryBoot.ZIP_DOMAIN, LibRepositoryBoot.ZIP_COMMENT_ATTRIBUTE);
    if (comment != null)
    {
      zipEntry.setComment(String.valueOf(comment));
    }
    final Object version = item.getAttribute(LibRepositoryBoot.REPOSITORY_DOMAIN, LibRepositoryBoot.VERSION_ATTRIBUTE);
    if (version instanceof Date)
    {
      final Date date = (Date) version;
      zipEntry.setTime(date.getTime());
    }

    final int zipMethod = RepositoryUtilities.getZipMethod(item);
    zipEntry.setCrc(crc32.getValue());
    if (zipMethod == Deflater.NO_COMPRESSION)
    {
      zipEntry.setCompressedSize(size);
      zipEntry.setSize(size);
    }
    else
    {
      zipEntry.setSize(size);
    }
    repository.writeContent(zipEntry, infi, zipMethod, RepositoryUtilities.getZipLevel(item));
    infi.close();

    closed = true;
    outputStream = null;
    deflaterOutputStream = null;
  }


  public void write(final byte[] b)
      throws IOException
  {
    if (closed)
    {
      throw new IOException("Already closed");
    }
    deflaterOutputStream.write(b);
    crc32.update(b);
    size += b.length;
  }

  public void flush()
      throws IOException
  {
    if (closed)
    {
      throw new IOException("Already closed");
    }
    deflaterOutputStream.flush();
  }
}


source/org/pentaho/reporting/libraries/repository/LibRepositoryInfo.java
/*
 * This program is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
 * Foundation.
 *
 * You should have received a copy of the GNU Lesser General Public License along with this
 * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
 * or from the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * Copyright (c) 2006 - 2009 Pentaho Corporation and Contributors.  All rights reserved.
 */

package org.pentaho.reporting.libraries.repository;

import org.pentaho.reporting.libraries.base.LibBaseInfo;
import org.pentaho.reporting.libraries.base.versioning.ProjectInformation;

/**
 * The LibRepositoryInfo class contains all dependency information and some
 * common information like version, license and contributors about the
 * library itself.
 *
 * @author Thomas Morgner
 */
public class LibRepositoryInfo extends ProjectInformation
{
  private static LibRepositoryInfo instance;

  /**
   * Returns the singleton instance of the ProjectInformation-class.
   *
   * @return the singleton ProjectInformation.
   */
  public static synchronized ProjectInformation getInstance()
  {
    if (instance == null)
    {
      instance = new LibRepositoryInfo();
      instance.initialize();
    }
    return instance;
  }

  /**
   * Constructs an empty project info object.
   */
  private LibRepositoryInfo()
  {
    super("librepository", "LibRepository");
  }

  /**
   * Initialized the project info object.
   */
  private void initialize()
  {
    setLicenseName("LGPL");

    setInfo("http://reporting.pentaho.org/librepository/");
    setCopyright("(C)opyright 2006-2010, by Pentaho Corporation and Contributors");

    setBootClass(LibRepositoryBoot.class.getName());
    addLibrary(LibBaseInfo.getInstance());
  }
}
